### PR TITLE
Port etag to master

### PR DESCRIPTION
**Pull request by Alexandru-Dumitru**
Addressing issue: https://github.com/zowe/zowe-cli/issues/596
Taking advantage of: https://github.com/zowe/imperative/pull/348
- added needed headers
- fix current unit tests for download and upload
- update index.ts to export new interfaces and types
- new interfaces created
- update download to use new imperative rest method and return etag if needed
- update upload to use new imperative rest method
    - pass and retrieve etag if needed
- deprecated several Upload uss functions in favor of ones that use IUploadOptions

### CODE REVIEW

**Comment by Alexandru-Dumitru** [comment]
> SonarCloud Quality Gate failed.
> 
> [Bug](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=VULNERABILITY) (and [Security Hotspot](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=SECURITY_HOTSPOT) to review)
> [Code Smell](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=zowe_zowe-cli&pullRequest=646&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=zowe_zowe-cli&pullRequest=646&metric=new_coverage&view=list) [58.8% Coverage](https://sonarcloud.io/component_measures?id=zowe_zowe-cli&pullRequest=646&metric=new_coverage&view=list)
> [](https://sonarcloud.io/component_measures?id=zowe_zowe-cli&pullRequest=646&metric=new_duplicated_lines_density&view=list) [8.3% Duplication](https://sonarcloud.io/component_measures?id=zowe_zowe-cli&pullRequest=646&metric=new_duplicated_lines_density&view=list)

will look at this tomorrow

**Comment by zFernand0** [comment]
Referencing a really cool comment on someone may test this functionality.
https://github.com/zowe/vscode-extension-for-zowe/issues/458#issuecomment-577109262

File: packages/zosfiles/src/api/methods/upload/Upload.ts [hunk 1]
```diff
@@ -551,19 +612,52 @@ export class Upload {
...
+     */
+    public static async fileToUSSFile(session: AbstractSession,
+                                      inputFile: string,
+                                      ussname: string,
+                                      binary: boolean = false,
+                                      localEncoding?: string,
+                                      task?: ITaskWithStatus,
+                                      etag?: string,
+                                      returnEtag?: boolean): Promise<IZosFilesResponse> {
+        return this.fileToUssFile(session, inputFile, ussname, {
```
[the remark this answers is not in the document]
Comment (Alexandru-Dumitru) [reply packages/zosfiles/src/api/methods/upload/Upload.ts]: I hope so as well 😆 . I couldn't come up with a better name, so I tried to stick as close as possible to the original.

File: packages/zosfiles/src/api/doc/types/ZosmfRestClientProperties.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+*/
+
+/**
+ * String type definition for properties of abstractRestClient that have a getter set.
+ * This can be safely used in a getter call as a variable for the abstractRestClient object.
+ * @export
+ * @typedef CLIENT_PROPERTY
+ */
+export type CLIENT_PROPERTY = "requestSuccess" | "requestFailure" | "data" | "dataString" | "response" | "session" | "log";
+export const CLIENT_PROPERTY = {
```
[the remark this answers is not in the document]
Comment (Alexandru-Dumitru) [reply packages/zosfiles/src/api/doc/types/ZosmfRestClientProperties.ts]: Yeah, I guess these should be exported by imperative and used here (talking about all interfaces that don't add new zosmf related options to existing ones). TBH I forgot why I re-created them. My assumption is that I forgot to export them from imperative, and directly importing from dependent package is forbidden (as it should)

File: packages/zosfiles/__tests__/api/methods/upload/Upload.unit.test.ts [hunk 3]
```diff
@@ -424,11 +487,262 @@ describe("z/OS Files - Upload", () => {
...
+            expect(error).toBeUndefined();
+            expect(response).toBeDefined();
+
+            expect(zosmfPutFullSpy).toHaveBeenCalledTimes(1);
+            expect(zosmfPutFullSpy).toHaveBeenCalledWith(dummySession, {resource:endpoint,
+                                                                        normalizeRequestNewLines: true,
+                                                                        reqHeaders,
+                                                                        requestStream: inputStream});
+        });
+        it("return with proper response when upload stream to a data set with optional parameters", async () => {
```
[the remark this answers is not in the document]
Comment (Alexandru-Dumitru) [reply packages/zosfiles/__tests__/api/methods/upload/Upload.unit.test.ts]: refactored 😄

File: packages/zosfiles/src/api/doc/types/ZosmfRestClientProperties.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zFernand0) [reply packages/zosfiles/src/api/doc/types/ZosmfRestClientProperties.ts]: We can tackle this in the `next` branches. See https://github.com/zowe/zowe-cli/issues/668

File: packages/zosfiles/src/api/doc/IRestClientResponse.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+*/
+
+import { Session, Logger } from "@zowe/imperative";
+
+/**
+ * Interface to map client's REST call response
+ * @export
+ * @interface IRestClientResponse
+ */
+export interface IRestClientResponse {
```
[the remark this answers is not in the document]
Comment (zFernand0) [reply packages/zosfiles/src/api/doc/IRestClientResponse.ts]: See https://github.com/zowe/zowe-cli/issues/668

**Pull request merged**